### ISSUE TITLE
Site down

### ISSUE THREAD

**Issue by cuppajoeman** [corollari/sharecaou#1]
I tried connected to sharecaou.com, but I timed out.

**Comment by corollari** [comment]
I took it down because of server costs :/

**Comment by cuppajoeman** [comment]
I see, thank you!

**Comment by corollari** [comment]
If you are interested in bringing it back up I'd love to help and I can also cover all server costs now.

**Comment by cuppajoeman** [comment]
Yeah, I'm actually really intrigued by this project! What could I do to help?

**Comment by corollari** [comment]
That's great! Is there any specific part you'd like to help with? Or just anything will do?

I must also say that I've neglected this project a bit, so if you'd like to become it's maintainer and drive it forward I could help from the sidelines :+1:

**Comment by cuppajoeman** [comment]
As of right now I don't know all the bits and pieces that are part of the project. So anything would do for me. 

I would love to work on this project, for right now I my priority are my online exams/classes, but I can start trying to understand the codebase (where should I start?) and when they are over I can do some more!

**Comment by corollari** [comment]
Allow me to give you a quick tour of the different parts then:
- [caou](https://github.com/corollari/caoutchouc): The main piece, the compiler that converts files to pdf. At the moment it's quite basic, as it only transpiles some parts into pandoc's format and then lets pandoc do the heavy lifting.
- [sharecaou](https://github.com/corollari/sharecaou): Docker image that changes sharelatex to use caou instead of the vanilla latex compiler. It works fine but is not scalable at all and is really hard to improve due to the fact that there is zero documentation on how sharelatex is structured.
- [caoutchouc-template](https://github.com/corollari/caoutchouc-template): Just a basic template that handles CI for projects using caoutchouc, used mostly by me to automatically generate pdfs of my documents.

Regarding where to start I think that working on caou would be best, as it could use a lot of improvements and the code is quite easy to grasp and improve.

**Comment by cuppajoeman** [comment]
Hi,

I have been reading through the index.js file (is this the right file to look through ?), here is my understanding and questions that came along.

- Lines 1 to 30 
  - deals with invalid arguments 
  - gets the filename and I'm not 100% sure but also is setting a custom typesetter. Is that right?
- Lines 32 to 55 
  - deal with pandoc installation and tries to compile using pandoc. How does the async part work?
- Lines 57 to 84 (compile function)
  - We remove replace the euro symbols (I don't see that key on my keyboard btw) with "$$asciimath", I'm assuming this is for when we call pandoc?
  - I'm a little confused by line 64 (removeUsePackage(input)), see the function at the bottom, but not sure too much what it does.
  - We start a child process running pandoc with flags as follow: (also I've just learned about child processes this semester, are you doing this so these two programs can run in parallel?)
    - ```-t latex``` specifies output format to be latex
    - ```-f markdown+lists_without_preceding_blankline+hard_line_breaks+raw_tex+raw_attribute``` I'm not sure about everything here, but seems to be setting the input format for what we type.
    - ```-s``` generates header and footer so it's standalone
    - ```--filter caou-pandoc-filter``` pushes it through caou-pandoc-filter before output (what is caou-pandoc-filter)
    - ```{input: input}``` I'm not sure what this last argument does.
    - does the .stdout return the result from this command? 
- lines 70 to 76 (I'm not too sure what the replacing does) 
- lines 78 to 83
  - Is it renaming the latex file to have .cautex extension?
  - Actually generate the pdf file using the typesetter they chose at the start
  - The following is a slice from the child process on line 83 ```process.argv.slice(2,-1).concat([filename+".cautex"]), { stdio: 'inherit' }```  What does the second argument to spawnSync represent, also the stdio: inherit ?

Ok, that was my quick look-over the index.js hopefully not too many questions, and just let me know if I am understanding it correctly... feel free to refer me to documentation if some questions would be better answered that way!

**Comment by corollari** [comment]
Yes, index.js is the main file and where everything happens.

> gets the filename and I'm not 100% sure but also is setting a custom typesetter. Is that right?

Correct, this part just handles arguments.

> deal with pandoc installation and tries to compile using pandoc. How does the async part work?

Correct, the async part is just there to use `await pandoc.run(['--version'])` in order to wait for the pandoc installation to finish, that whole part could be rewritten in the following way (lines 46-54):
```
pandoc.run(['--version'])
.then(()=>{
	compile(input, typesetter, inputFile, pandoc.path())
}).catch((e=>{
	console.error('✗ pandoc installation failed')
	return 1
}))
```

> We remove replace the euro symbols (I don't see that key on my keyboard btw) with "$$asciimath", I'm assuming this is for when we call pandoc?

This is done in order to move all ascimath into code blocks which can be processed inside pandoc with a filter (a name used for pandoc add-ons).

> I'm a little confused by line 64 (removeUsePackage(input)), see the function at the bottom, but not sure too much what it does.

It just removes lines like `\usepackage{...}` before handing the document to pandoc, this lines will be readded later, this is done because pandoc can't handle this lines properly.

> are you doing this so these two programs can run in parallel?

No, here we are running the processes sequentially not in parallel, the reason we are creating a new process is because that's the easiest way to compile with pandoc from node.

> I'm not sure about everything here, but seems to be setting the input format for what we type.

Yep

> pushes it through caou-pandoc-filter before output (what is caou-pandoc-filter)

caou-pandoc-filter is the pandoc add-on I mentioned before which is used to process asciimath, essentially the way it works is that pandoc does some initially processing, then runs caou-pandoc-filter on the pieces of code we marked before with `$$asciimath` and then continues processing after-wards. The code of caou-pandoc-filter is on the file `pandoc-filter.js`, it just compiles asciimath to latex.

> {input: input} I'm not sure what this last argument does.

Just sends the current code as input for pandoc

> does the .stdout return the result from this command?

Yes

> lines 70 to 76 (I'm not too sure what the replacing does)

It's replacing the `$$asciimath` we added before (so pandoc could process them with our special filter) for `$` so latex can process it. Also, it adds the \usepackage{} statements back into the file.

> Is it renaming the latex file to have .cautex extension?

Kinda, it takes the latex file we just generated and stores it as a cautex file.

> What does the second argument to spawnSync represent?

It calls the typesetter with all the extra arguments that you have provided to `caou` when calling it.

>  What does the second argument to spawnSync represent, also the stdio: inherit ?

It just pipes all the output of the latex typesetter to the shell so the user running `caou` can see it.

----

 Let me know if you have any other questions :+1:

**Comment by cuppajoeman** [comment]
When I run the program without the `raw_attribute` (in the line where we call pandoc) the newest version of pandoc runs fine. 

So we are using `raw_tex` which allows us to write latex directly without having to do \`\`\`{=latex} ... , but we are also using `raw_attribute` which means we have to use \`\`\`{=latex} ... So maybe we just don't need `raw_attribute`. I could be wrong about this, though maybe they can be used in conjunction.

I did a little more testing and I found the reason why it's breaking on the newest version of pandoc.

![image](https://user-images.githubusercontent.com/19198104/79705488-4271ad80-828c-11ea-8409-899d2d045226.png)

(when `raw_attribute` is removed, it works perfectly)

I've found the problem, my pandoc is defaulting to version 1.19.2.4 which isn't the newest version, though it is the one that gets downloaded when I do ```sudo apt install pandoc```

**Comment by corollari** [comment]
Tomorrow I'll look into why we are using `raw_attribute` (I don't remember the reason right now). Also sorry for taking a long time to answer, I have been moving to a different country because of coronavirus.

**Comment by cuppajoeman** [comment]
No worries, this is a difficult time. So take all the time you need !

**Comment by corollari** [comment]
After looking more into it, I have to agree with you, `raw_attribute` is not needed. I've run the program without `raw_attribute` through all the current tests and there doesn't seem to be any change, and by looking at the code I don't see any reason that would lead me to believe that it is needed.

I think the reason why it is included is that, in one of the initial iterations of caou, I used formatted code blocks instead of Math blocks to contain asciimath, along with a mechanism that used `raw_attribute`, and I didn't remove that plugin when I changed the mechanism.
